Add Nvidia/cuda compilers to CI

**What Needs to be Done?**
<!--- What needs to happen and why?  How does it relate to the rest of the  -->
<!--- project?  -->
Add a new container with CUDA compilers for testing VT.

There is a failure much later from checkpoint:

```
2020-07-07T08:03:43.3550814Z [250/441] Building CXX object tests/CMakeFiles/test_rdma_handle.dir/unit/rdma/test_rdma_handle.cc.o
2020-07-07T08:03:43.3551079Z FAILED: tests/CMakeFiles/test_rdma_handle.dir/unit/rdma/test_rdma_handle.cc.o 
2020-07-07T08:03:43.3552849Z /usr/bin/ccache /nvcc_wrapper/build/nvcc_wrapper  -DFMT_HEADER_ONLY=1 -DFMT_USE_USER_DEFINED_LITERALS=0 -DHAS_DETECTION_COMPONENT=1 -I/vt/tests/unit -isystem /vt/tests/extern/googletest/googletest/include -I/vt/lib/fmt -I/vt/lib/CLI -isystem /vt/tests/extern/googletest/googletest -Irelease -I/vt/src -isystem /usr/local/include -isystem /build/checkpoint/install/include -isystem /build/detector/install/include -Wall -pedantic -Wshadow -Wno-unknown-pragmas -Wsign-compare -O3 -DNDEBUG   -fdiagnostics-color=always -std=c++1y -fPIC -fopenmp -std=c++14 -MD -MT tests/CMakeFiles/test_rdma_handle.dir/unit/rdma/test_rdma_handle.cc.o -MF tests/CMakeFiles/test_rdma_handle.dir/unit/rdma/test_rdma_handle.cc.o.d -o tests/CMakeFiles/test_rdma_handle.dir/unit/rdma/test_rdma_handle.cc.o -c /vt/tests/unit/rdma/test_rdma_handle.cc
2020-07-07T08:03:43.3553672Z nvcc_wrapper does not accept standard flags -std=c++1y since partial standard flags and standards after C++14 are not supported. nvcc_wrapper will use -std=c++14 instead. It is undefined behavior to use this flag. This should only be occurring during CMake configuration.
2020-07-07T08:03:43.3554245Z nvcc_wrapper - *warning* you have set multiple standard flags (-std=c++1* or --std=c++1*), only the last is used because nvcc can only accept a single std setting
2020-07-07T08:03:43.3555795Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h: In instantiation of 'void checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::apply(SerializerT&, T*, checkpoint::SerialSizeType, checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::hasNotVirtualSerialize<U>*) [with U = vt::rdma::impl::HandleData; SerializerT = checkpoint::Sizer; T = vt::rdma::impl::HandleData; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, vt::rdma::impl::HandleData>; checkpoint::SerialSizeType = long unsigned int; checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::hasNotVirtualSerialize<U> = vt::rdma::impl::HandleData]':
2020-07-07T08:03:43.3557352Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:99:13:   required from 'void checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::operator()(SerializerT&, T*, checkpoint::SerialSizeType) [with U = vt::rdma::impl::HandleData; SerializerT = checkpoint::Sizer; T = vt::rdma::impl::HandleData; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, vt::rdma::impl::HandleData>; checkpoint::SerialSizeType = long unsigned int]'
2020-07-07T08:03:43.3558853Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_byte.h:107:9:   required from 'void checkpoint::dispatch::SerializerDispatchByte<SerializerT, T, Dispatcher>::operator()(SerializerT&, T*, checkpoint::SerialSizeType, checkpoint::dispatch::SerializerDispatchByte<SerializerT, T, Dispatcher>::isNotByteCopyType<U>*) [with U = vt::rdma::impl::HandleData; SerializerT = checkpoint::Sizer; T = vt::rdma::impl::HandleData; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, vt::rdma::impl::HandleData>; checkpoint::SerialSizeType = long unsigned int; checkpoint::dispatch::SerializerDispatchByte<SerializerT, T, Dispatcher>::isNotByteCopyType<U> = vt::rdma::impl::HandleData]'
2020-07-07T08:03:43.3560017Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch.impl.h:71:3:   required from 'static TraverserT& checkpoint::dispatch::Traverse::with(T&, TraverserT&, checkpoint::SerialSizeType) [with T = vt::rdma::impl::HandleData; TraverserT = checkpoint::Sizer; checkpoint::SerialSizeType = long unsigned int]'
2020-07-07T08:03:43.3560890Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch.impl.h:55:32:   required from 'Serializer& checkpoint::operator|(Serializer&, T&) [with Serializer = checkpoint::Sizer; T = vt::rdma::impl::HandleData]'
2020-07-07T08:03:43.3561728Z /vt/src/vt/collective/reduce/operators/default_msg.h:95:3:   required from 'void vt::collective::reduce::operators::ReduceDataMsg<DataType>::serialize(SerializeT&) [with SerializeT = checkpoint::Sizer; DataType = vt::rdma::impl::HandleData]'
2020-07-07T08:03:43.3562293Z /vt/src/vt/messaging/message/message_serialize.h:396:16:   [ skipping 10 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
2020-07-07T08:03:43.3563727Z /vt/src/vt/collective/reduce/operators/default_op.impl.h:64:1:   required from 'static void vt::collective::reduce::operators::ReduceCombine<T>::msgHandler(MsgT*) [with MsgT = vt::rdma::impl::ConstructMsg<short unsigned int, (vt::rdma::HandleEnum)1, vt::objgroup::proxy::Proxy<vt::tests::unit::TestObjGroup> >; Op = vt::collective::reduce::operators::PlusOp<vt::rdma::impl::HandleData>; ActOp = vt::collective::reduce::operators::ReduceCallback<vt::rdma::impl::ConstructMsg<short unsigned int, (vt::rdma::HandleEnum)1, vt::objgroup::proxy::Proxy<vt::tests::unit::TestObjGroup> > >; T = void]'
2020-07-07T08:03:43.3565836Z /vt/src/vt/objgroup/proxy/proxy_objgroup.h:134:21:   required by substitution of 'template<class OpT, class MsgPtrT, class MsgT, void (* f)(MsgT*)> void vt::objgroup::proxy::Proxy<vt::rdma::Manager>::reduce<OpT, MsgPtrT, MsgT, f>(MsgPtrT, vt::Callback<MsgT>, vt::objgroup::proxy::Proxy<vt::rdma::Manager>::ReduceStamp) const [with OpT = vt::collective::reduce::operators::PlusOp<vt::rdma::impl::HandleData>; MsgPtrT = vt::rdma::impl::ConstructMsg<short unsigned int, (vt::rdma::HandleEnum)1, vt::objgroup::proxy::Proxy<vt::tests::unit::TestObjGroup> >*; MsgT = vt::rdma::impl::ConstructMsg<short unsigned int, (vt::rdma::HandleEnum)1, vt::objgroup::proxy::Proxy<vt::tests::unit::TestObjGroup> >; void (* f)(MsgT*) = vt::collective::reduce::operators::ReduceCombine<>::msgHandler<vt::rdma::impl::ConstructMsg<short unsigned int, (vt::rdma::HandleEnum)1, vt::objgroup::proxy::Proxy<vt::tests::unit::TestObjGroup> >, vt::collective::reduce::operators::PlusOp<vt::rdma::impl::HandleData>, vt::collective::reduce::operators::ReduceCallback<vt::rdma::impl::ConstructMsg<short unsigned int, (vt::rdma::HandleEnum)1, vt::objgroup::proxy::Proxy<vt::tests::unit::TestObjGroup> > > >]'
2020-07-07T08:03:43.3567104Z /vt/src/vt/rdmahandle/manager.impl.h:139:1:   required from 'vt::rdma::Handle<T, E> vt::rdma::Manager::makeHandleCollectiveObjGroup(ProxyT, std::size_t, bool) [with T = short unsigned int; vt::rdma::HandleEnum E = (vt::rdma::HandleEnum)1; ProxyT = vt::objgroup::proxy::Proxy<vt::tests::unit::TestObjGroup>; std::size_t = long unsigned int]'
2020-07-07T08:03:43.3567993Z /vt/src/vt/objgroup/proxy/proxy_objgroup.impl.h:169:128:   required from 'vt::rdma::Handle<T> vt::objgroup::proxy::Proxy<ObjT>::makeHandleRDMA(std::size_t, bool) const [with T = short unsigned int; ObjT = vt::tests::unit::TestObjGroup; std::size_t = long unsigned int]'
2020-07-07T08:03:43.3568938Z /vt/tests/unit/rdma/test_rdma_handle.cc:64:61:   required from 'vt::HandleRDMA<T, short int> vt::tests::unit::TestObjGroup::makeHandle(std::size_t, bool) [with T = short unsigned int; vt::HandleRDMA<T, short int> = vt::rdma::Handle<short unsigned int, (vt::rdma::HandleEnum)1, short int, void>; std::size_t = long unsigned int]'
2020-07-07T08:03:43.3569818Z /vt/tests/unit/rdma/test_rdma_handle.cc:245:50:   required from 'void vt::tests::unit::gtest_suite_TestRDMAHandle_::test_rdma_handle_5<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = short unsigned int]'
2020-07-07T08:03:43.3570051Z /vt/src/vt/group/region/group_shallow_list.h:55:8:   required from here
2020-07-07T08:03:43.3571117Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:218:19: error: no matching function for call to 'checkpoint::dispatch::SerializerDispatchNonByte<checkpoint::Sizer, vt::rdma::impl::HandleData, checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, vt::rdma::impl::HandleData> >::applyStatic(checkpoint::Sizer&, vt::rdma::impl::HandleData*&, checkpoint::SerialSizeType&)'
2020-07-07T08:03:43.3571377Z      return applyStatic(s, val, num);
2020-07-07T08:03:43.3571533Z         ~~~~~~~~~~~^~~~~~~~~~~~~
2020-07-07T08:03:43.3571870Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:163:1: note: candidate: template<class U> void checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::applyStatic(SerializerT&, T*, checkpoint::SerialSizeType, checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::hasInSerialize<U>*) [with U = U; SerializerT = checkpoint::Sizer; T = vt::rdma::impl::HandleData; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, vt::rdma::impl::HandleData>]
2020-07-07T08:03:43.3572168Z    void applyStatic(
2020-07-07T08:03:43.3572310Z  ^ ~~~~~~~~~
2020-07-07T08:03:43.3572502Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:163:1: note:   template argument deduction/substitution failed:
2020-07-07T08:03:43.3572859Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:177:1: note: candidate: template<class U> void checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::applyStatic(SerializerT&, T*, checkpoint::SerialSizeType, checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::hasNoninSerialize<U>*) [with U = U; SerializerT = checkpoint::Sizer; T = vt::rdma::impl::HandleData; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, vt::rdma::impl::HandleData>]
2020-07-07T08:03:43.3573159Z    void applyStatic(
2020-07-07T08:03:43.3573297Z  ^ ~~~~~~~~~
2020-07-07T08:03:43.3647914Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:177:1: note:   template argument deduction/substitution failed:
2020-07-07T08:03:43.3648322Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:190:1: note: candidate: template<class U> void checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::applyStatic(SerializerT&, T*, checkpoint::SerialSizeType, checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::isEnum<U>*) [with U = U; SerializerT = checkpoint::Sizer; T = vt::rdma::impl::HandleData; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, vt::rdma::impl::HandleData>]
2020-07-07T08:03:43.3648816Z    void applyStatic(
2020-07-07T08:03:43.3654477Z  ^ ~~~~~~~~~
2020-07-07T08:03:43.3654760Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:190:1: note:   template argument deduction/substitution failed:
2020-07-07T08:03:43.3655695Z /build/checkpoint/install/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:218:31: error: return-statement with a value, in function returning 'void' [-fpermissive]
2020-07-07T08:03:43.3655911Z      return applyStatic(s, val, num);
2020-07-07T08:03:43.3656091Z 
```

--- a/.github/workflows/dockerimage-intel-18-ubuntu-mpich-extended.yml
+++ b/.github/workflows/dockerimage-intel-18-ubuntu-mpich-extended.yml
@@ -11,7 +11,7 @@ on:
 
 jobs:
 
-  build:
+  build_optional:
 
     if: github.event.label.name == 'runextendedtests'
     runs-on: ubuntu-latest

--- a/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich-extended.yml
+++ b/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich-extended.yml
@@ -1,0 +1,83 @@
+###############################################################################
+############## Warning this is a generated file---do not modify ###############
+###############################################################################
+
+name: PR tests extended (nvidia cuda 10.1, ubuntu, mpich)
+
+on:
+  pull_request:
+    types:
+      - labeled
+
+jobs:
+
+  build:
+
+    if: github.event.label.name == 'runextendedtests'
+    runs-on: ubuntu-latest
+
+    env:
+      REPO: lifflander1/vt
+      ARCH: amd64
+      UBUNTU: 18.04
+      COMPILER_TYPE: nvidia
+      COMPILER: nvcc-10
+      BUILD_TYPE: release
+      ULIMIT_CORE: 0
+      CODE_COVERAGE: 0
+      VT_LB: 1
+      VT_TRACE: 1
+      VT_TRACE_RT: 0
+      VT_MIMALLOC: 0
+      VT_DOCS: 0
+      VT_ASAN: 0
+      VT_POOL: 0
+      VT_EXTENDED_TESTS: 1
+      VT_ZOLTAN: 0
+      CACHE: ~/.local/cache/
+
+    steps:
+    - name: Setup Build Root
+      uses: allenevans/set-env@v1.0.0
+      with:
+        BUILD_ROOT: "~/.local/cache/${{ env.ARCH }}-ubuntu-${{ env.UBUNTU }}-${{ env.COMPILER }}-cache/"
+    - name: Prepare caching timestamp
+      id: cache_ts
+      shell: cmake -P {0}
+      run: |
+        string(TIMESTAMP current_date "%Y-%m-%d-%H;%M;%S" UTC)
+        message("::set-output name=timestamp::${current_date}")
+    - uses: actions/cache@v1
+      env:
+        cache-name: ubuntu-nvidia-10-cache
+      with:
+        path: ${{ env.BUILD_ROOT }}/ccache
+        key: ${{ runner.os }}-${{ env.cache-name }}-${{ steps.cache_ts.outputs.timestamp }}
+        restore-keys: |
+          ${{ runner.os }}-${{ env.cache-name }}-
+    - uses: actions/checkout@v2
+    - name: Docker Pull Base Image
+      shell: bash
+      run: docker-compose pull --ignore-pull-failures ubuntu-cpp-clean
+    - name: Build the Docker image
+      run: docker-compose run ubuntu-cpp-clean
+    - name: Docker Push Base Image
+      if: success()
+      continue-on-error: true
+      shell: bash
+      run: |
+        docker login -u ${{ secrets.DOCKER_USERNAME }} \
+                     -p ${{ secrets.DOCKER_PASSWORD }}
+        docker-compose push ubuntu-cpp-clean
+    - name: Zip up CMake output
+      run: |
+        zip -j LastTest.log.gz ${{ env.BUILD_ROOT }}/vt/Testing/Temporary/LastTest.log
+        zip -j cmake-output.log.gz ${{ env.BUILD_ROOT }}/vt/cmake-output.log
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake test output
+        path: cmake-output.log.gz
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake full output
+        path: LastTest.log.gz

--- a/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich-extended.yml
+++ b/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich-extended.yml
@@ -11,7 +11,7 @@ on:
 
 jobs:
 
-  build:
+  build_optional:
 
     if: github.event.label.name == 'runextendedtests'
     runs-on: ubuntu-latest

--- a/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
+++ b/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
@@ -13,7 +13,7 @@ on:
 
 jobs:
 
-  build:
+  build_optional:
 
 
     runs-on: ubuntu-latest

--- a/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
+++ b/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
@@ -1,0 +1,77 @@
+name: PR tests (nvidia cuda 10.1, ubuntu, mpich)
+
+on:
+  push:
+    branches:
+      - develop
+      - 1.*
+  pull_request:
+
+jobs:
+
+  build:
+
+    runs-on: ubuntu-latest
+
+    env:
+      REPO: lifflander1/vt
+      ARCH: amd64
+      UBUNTU: 18.04
+      COMPILER_TYPE: nvidia
+      COMPILER: nvcc-10
+      BUILD_TYPE: release
+      ULIMIT_CORE: 0
+      VT_LB: 1
+      VT_TRACE: 1
+      VT_TRACE_RT: 0
+      VT_MIMALLOC: 0
+      VT_DOCS: 0
+      VT_ASAN: 0
+      VT_POOL: 0
+      CACHE: ~/.local/cache/
+
+    steps:
+    - name: Setup Build Root
+      uses: allenevans/set-env@v1.0.0
+      with:
+        BUILD_ROOT: "~/.local/cache/${{ env.ARCH }}-ubuntu-${{ env.UBUNTU }}-${{ env.COMPILER }}-cache/"
+    - name: Prepare caching timestamp
+      id: cache_ts
+      shell: cmake -P {0}
+      run: |
+        string(TIMESTAMP current_date "%Y-%m-%d-%H;%M;%S" UTC)
+        message("::set-output name=timestamp::${current_date}")
+    - uses: actions/cache@v1
+      env:
+        cache-name: ubuntu-nvidia-10-cache
+      with:
+        path: ${{ env.BUILD_ROOT }}/ccache
+        key: ${{ runner.os }}-${{ env.cache-name }}-${{ steps.cache_ts.outputs.timestamp }}
+        restore-keys: |
+          ${{ runner.os }}-${{ env.cache-name }}-
+    - uses: actions/checkout@v2
+    - name: Docker Pull Base Image
+      shell: bash
+      run: docker-compose pull --ignore-pull-failures ubuntu-cpp-clean
+    - name: Build the Docker image
+      run: docker-compose run ubuntu-cpp-clean
+    - name: Docker Push Base Image
+      if: success()
+      continue-on-error: true
+      shell: bash
+      run: |
+        docker login -u ${{ secrets.DOCKER_USERNAME }} \
+                     -p ${{ secrets.DOCKER_PASSWORD }}
+        docker-compose push ubuntu-cpp-clean
+    - name: Zip up CMake output
+      run: |
+        zip -j LastTest.log.gz ${{ env.BUILD_ROOT }}/vt/Testing/Temporary/LastTest.log
+        zip -j cmake-output.log.gz ${{ env.BUILD_ROOT }}/vt/cmake-output.log
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake test output
+        path: cmake-output.log.gz
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake full output
+        path: LastTest.log.gz

--- a/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
+++ b/.github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
@@ -1,3 +1,7 @@
+###############################################################################
+############## Warning this is a generated file---do not modify ###############
+###############################################################################
+
 name: PR tests (nvidia cuda 10.1, ubuntu, mpich)
 
 on:
@@ -11,6 +15,7 @@ jobs:
 
   build:
 
+
     runs-on: ubuntu-latest
 
     env:
@@ -21,6 +26,7 @@ jobs:
       COMPILER: nvcc-10
       BUILD_TYPE: release
       ULIMIT_CORE: 0
+      CODE_COVERAGE: 0
       VT_LB: 1
       VT_TRACE: 1
       VT_TRACE_RT: 0
@@ -28,6 +34,8 @@ jobs:
       VT_DOCS: 0
       VT_ASAN: 0
       VT_POOL: 0
+      VT_EXTENDED_TESTS: ${{ github.event_name != 'pull_request' }}
+      VT_ZOLTAN: 0
       CACHE: ~/.local/cache/
 
     steps:

--- a/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich-extended.yml
+++ b/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich-extended.yml
@@ -11,7 +11,7 @@ on:
 
 jobs:
 
-  build:
+  build_optional:
 
     if: github.event.label.name == 'runextendedtests'
     runs-on: ubuntu-latest

--- a/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich-extended.yml
+++ b/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich-extended.yml
@@ -1,0 +1,83 @@
+###############################################################################
+############## Warning this is a generated file---do not modify ###############
+###############################################################################
+
+name: PR tests extended (nvidia cuda 11.0, ubuntu, mpich)
+
+on:
+  pull_request:
+    types:
+      - labeled
+
+jobs:
+
+  build:
+
+    if: github.event.label.name == 'runextendedtests'
+    runs-on: ubuntu-latest
+
+    env:
+      REPO: lifflander1/vt
+      ARCH: amd64
+      UBUNTU: 18.04
+      COMPILER_TYPE: nvidia
+      COMPILER: nvcc-11
+      BUILD_TYPE: release
+      ULIMIT_CORE: 0
+      CODE_COVERAGE: 0
+      VT_LB: 1
+      VT_TRACE: 1
+      VT_TRACE_RT: 0
+      VT_MIMALLOC: 0
+      VT_DOCS: 0
+      VT_ASAN: 0
+      VT_POOL: 0
+      VT_EXTENDED_TESTS: 1
+      VT_ZOLTAN: 0
+      CACHE: ~/.local/cache/
+
+    steps:
+    - name: Setup Build Root
+      uses: allenevans/set-env@v1.0.0
+      with:
+        BUILD_ROOT: "~/.local/cache/${{ env.ARCH }}-ubuntu-${{ env.UBUNTU }}-${{ env.COMPILER }}-cache/"
+    - name: Prepare caching timestamp
+      id: cache_ts
+      shell: cmake -P {0}
+      run: |
+        string(TIMESTAMP current_date "%Y-%m-%d-%H;%M;%S" UTC)
+        message("::set-output name=timestamp::${current_date}")
+    - uses: actions/cache@v1
+      env:
+        cache-name: ubuntu-nvidia-11-cache
+      with:
+        path: ${{ env.BUILD_ROOT }}/ccache
+        key: ${{ runner.os }}-${{ env.cache-name }}-${{ steps.cache_ts.outputs.timestamp }}
+        restore-keys: |
+          ${{ runner.os }}-${{ env.cache-name }}-
+    - uses: actions/checkout@v2
+    - name: Docker Pull Base Image
+      shell: bash
+      run: docker-compose pull --ignore-pull-failures ubuntu-cpp-clean
+    - name: Build the Docker image
+      run: docker-compose run ubuntu-cpp-clean
+    - name: Docker Push Base Image
+      if: success()
+      continue-on-error: true
+      shell: bash
+      run: |
+        docker login -u ${{ secrets.DOCKER_USERNAME }} \
+                     -p ${{ secrets.DOCKER_PASSWORD }}
+        docker-compose push ubuntu-cpp-clean
+    - name: Zip up CMake output
+      run: |
+        zip -j LastTest.log.gz ${{ env.BUILD_ROOT }}/vt/Testing/Temporary/LastTest.log
+        zip -j cmake-output.log.gz ${{ env.BUILD_ROOT }}/vt/cmake-output.log
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake test output
+        path: cmake-output.log.gz
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake full output
+        path: LastTest.log.gz

--- a/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
+++ b/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
@@ -13,7 +13,7 @@ on:
 
 jobs:
 
-  build:
+  build_optional:
 
 
     runs-on: ubuntu-latest

--- a/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
+++ b/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
@@ -1,3 +1,7 @@
+###############################################################################
+############## Warning this is a generated file---do not modify ###############
+###############################################################################
+
 name: PR tests (nvidia cuda 11.0, ubuntu, mpich)
 
 on:
@@ -11,6 +15,7 @@ jobs:
 
   build:
 
+
     runs-on: ubuntu-latest
 
     env:
@@ -21,6 +26,7 @@ jobs:
       COMPILER: nvcc-11
       BUILD_TYPE: release
       ULIMIT_CORE: 0
+      CODE_COVERAGE: 0
       VT_LB: 1
       VT_TRACE: 1
       VT_TRACE_RT: 0
@@ -28,6 +34,8 @@ jobs:
       VT_DOCS: 0
       VT_ASAN: 0
       VT_POOL: 0
+      VT_EXTENDED_TESTS: ${{ github.event_name != 'pull_request' }}
+      VT_ZOLTAN: 0
       CACHE: ~/.local/cache/
 
     steps:

--- a/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
+++ b/.github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
@@ -1,0 +1,77 @@
+name: PR tests (nvidia cuda 11.0, ubuntu, mpich)
+
+on:
+  push:
+    branches:
+      - develop
+      - 1.*
+  pull_request:
+
+jobs:
+
+  build:
+
+    runs-on: ubuntu-latest
+
+    env:
+      REPO: lifflander1/vt
+      ARCH: amd64
+      UBUNTU: 18.04
+      COMPILER_TYPE: nvidia
+      COMPILER: nvcc-11
+      BUILD_TYPE: release
+      ULIMIT_CORE: 0
+      VT_LB: 1
+      VT_TRACE: 1
+      VT_TRACE_RT: 0
+      VT_MIMALLOC: 0
+      VT_DOCS: 0
+      VT_ASAN: 0
+      VT_POOL: 0
+      CACHE: ~/.local/cache/
+
+    steps:
+    - name: Setup Build Root
+      uses: allenevans/set-env@v1.0.0
+      with:
+        BUILD_ROOT: "~/.local/cache/${{ env.ARCH }}-ubuntu-${{ env.UBUNTU }}-${{ env.COMPILER }}-cache/"
+    - name: Prepare caching timestamp
+      id: cache_ts
+      shell: cmake -P {0}
+      run: |
+        string(TIMESTAMP current_date "%Y-%m-%d-%H;%M;%S" UTC)
+        message("::set-output name=timestamp::${current_date}")
+    - uses: actions/cache@v1
+      env:
+        cache-name: ubuntu-nvidia-11-cache
+      with:
+        path: ${{ env.BUILD_ROOT }}/ccache
+        key: ${{ runner.os }}-${{ env.cache-name }}-${{ steps.cache_ts.outputs.timestamp }}
+        restore-keys: |
+          ${{ runner.os }}-${{ env.cache-name }}-
+    - uses: actions/checkout@v2
+    - name: Docker Pull Base Image
+      shell: bash
+      run: docker-compose pull --ignore-pull-failures ubuntu-cpp-clean
+    - name: Build the Docker image
+      run: docker-compose run ubuntu-cpp-clean
+    - name: Docker Push Base Image
+      if: success()
+      continue-on-error: true
+      shell: bash
+      run: |
+        docker login -u ${{ secrets.DOCKER_USERNAME }} \
+                     -p ${{ secrets.DOCKER_PASSWORD }}
+        docker-compose push ubuntu-cpp-clean
+    - name: Zip up CMake output
+      run: |
+        zip -j LastTest.log.gz ${{ env.BUILD_ROOT }}/vt/Testing/Temporary/LastTest.log
+        zip -j cmake-output.log.gz ${{ env.BUILD_ROOT }}/vt/cmake-output.log
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake test output
+        path: cmake-output.log.gz
+    - uses: actions/upload-artifact@v1
+      with:
+        name: CMake full output
+        path: LastTest.log.gz

--- a/README.md
+++ b/README.md
@@ -14,6 +14,8 @@
 ![clang-5, ubuntu, mpich, trace](https://github.com/DARMA-tasking/vt/workflows/PR%20tests%20(clang-5.0,%20ubuntu,%20mpich,%20trace)/badge.svg?branch=develop)
 ![apple clang, macosx, mpich](https://github.com/DARMA-tasking/vt/workflows/PR%20tests%20(clang-8,%20macosx,%20mpich)/badge.svg?branch=develop)
 ![intel 18.03, ubuntu, mpich](https://github.com/DARMA-tasking/vt/workflows/PR%20tests%20(intel%2018.03,%20ubuntu,%20mpich)/badge.svg?branch=develop)
+![nvidia cuda 10.1, ubuntu, mpich](https://github.com/DARMA-tasking/vt/workflows/PR%20tests%20(nvidia%20cuda%2010.1,%20ubuntu,%20mpich)/badge.svg?branch=develop)
+![nvidia cuda 11.0, ubuntu, mpich](https://github.com/DARMA-tasking/vt/workflows/PR%20tests%20(nvidia%20cuda%2011.0,%20ubuntu,%20mpich)/badge.svg?branch=develop)
 ![Build Documentation](https://github.com/DARMA-tasking/vt/workflows/Build%20Documentation/badge.svg?branch=develop)
 
 ## Introduction : What is *vt*?

--- a/ci/build_cpp.sh
+++ b/ci/build_cpp.sh
@@ -5,6 +5,10 @@ set -ex
 source_dir=${1}
 build_dir=${2}
 
+# Dependency versions, when fetched via git.
+detector_rev=master
+checkpoint_rev=develop
+
 if test "${VT_DOXYGEN_ENABLED:-0}" -eq 1
 then
     token=${3}
@@ -32,7 +36,7 @@ if test -d "${source_dir}/lib/detector"
 then
     echo "Detector already in lib... not downloading, building, and installing"
 else
-    git clone -b master --depth 1 https://github.com/DARMA-tasking/detector.git
+    git clone -b "${detector_rev}" --depth 1 https://github.com/DARMA-tasking/detector.git
     export DETECTOR=$PWD/detector
     export DETECTOR_BUILD=${build_dir}/detector
     mkdir -p "$DETECTOR_BUILD"
@@ -49,7 +53,7 @@ if test -d "${source_dir}/lib/checkpoint"
 then
     echo "Checkpoint already in lib... not downloading, building, and installing"
 else
-    git clone -b develop --depth 1 https://github.com/DARMA-tasking/checkpoint.git
+    git clone -b "${checkpoint_rev}" --depth 1 https://github.com/DARMA-tasking/checkpoint.git
     export CHECKPOINT=$PWD/checkpoint
     export CHECKPOINT_BUILD=${build_dir}/checkpoint
     mkdir -p "$CHECKPOINT_BUILD"

--- a/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
+++ b/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
@@ -69,7 +69,7 @@ RUN git clone https://github.com/kokkos/nvcc_wrapper.git && \
     cd nvcc_wrapper && \
     mkdir build && \
     cd build && \
-    cmake ../ \
+    cmake ../
 
 ENV MPI_EXTRA_FLAGS="" \
     CXX=/nvcc_wrapper/build/nvcc_wrapper \

--- a/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
+++ b/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
@@ -1,9 +1,10 @@
 
 ARG cuda=10.1
-FROM nvidia/cuda:${cuda}-devel-ubuntu18.04 as base
+ARG arch=amd64
+FROM ${arch}/ubuntu:18.04 as base
 
 ARG proxy=""
-ARG compiler=gcc-7
+ARG compiler=nvcc-10
 
 ENV https_proxy=${proxy} \
     http_proxy=${proxy}
@@ -13,25 +14,65 @@ ENV DEBIAN_FRONTEND=noninteractive
 RUN apt-get update -y -q && \
     apt-get install -y -q --no-install-recommends \
     ca-certificates \
+    g++-7 \
     curl \
     cmake \
+    less \
     git \
-    mpich \
-    libmpich-dev \
     wget \
-    ${compiler} \
     zlib1g \
     zlib1g-dev \
     ninja-build \
+    gnupg \
+    make-guile \
+    libomp5 \
     valgrind \
     ccache && \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/*
 
+RUN wget http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-ubuntu1804.pin && \
+    mv cuda-ubuntu1804.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
+    wget http://developer.download.nvidia.com/compute/cuda/10.1/Prod/local_installers/cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb && \
+    dpkg -i cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb && \
+    apt-key add /var/cuda-repo-10-1-local-10.1.243-418.87.00/7fa2af80.pub && \
+    apt-get update && \
+    apt-get -y install cuda && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* && \
+    rm -rf cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb
+
+ENV CC=gcc \
+    CXX=g++
+
+RUN wget http://www.mpich.org/static/downloads/3.3.2/mpich-3.3.2.tar.gz && \
+    tar xzf mpich-3.3.2.tar.gz && \
+    rm mpich-3.3.2.tar.gz && \
+    cd mpich-3.3.2 && \
+    ./configure \
+        --enable-static=false \
+        --enable-alloca=true \
+        --disable-long-double \
+        --enable-threads=single \
+        --enable-fortran=no \
+        --enable-fast=all \
+        --enable-g=none \
+        --enable-timing=none && \
+    make -j4 && \
+    make install && \
+    cd - && \
+    rm -rf mpich-3.3.2
+
+ENV CUDACXX=/usr/local/cuda-10.1/bin/nvcc
+
+RUN git clone https://github.com/kokkos/nvcc_wrapper.git && \
+    cd nvcc_wrapper && \
+    mkdir build && \
+    cd build && \
+    cmake ../ \
+
 ENV MPI_EXTRA_FLAGS="" \
-    CMAKE_PREFIX_PATH="/lib/x86_64-linux-gnu/" \
-    CC=nvcc \
-    CXX=nvcc \
+    CXX=/nvcc_wrapper/build/nvcc_wrapper \
     PATH=/usr/lib/ccache/:$PATH
 
 FROM base as build

--- a/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
+++ b/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
@@ -31,16 +31,31 @@ RUN apt-get update -y -q && \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/*
 
-RUN wget http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-ubuntu1804.pin && \
-    mv cuda-ubuntu1804.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
-    wget http://developer.download.nvidia.com/compute/cuda/10.1/Prod/local_installers/cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb && \
-    dpkg -i cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb && \
-    apt-key add /var/cuda-repo-10-1-local-10.1.243-418.87.00/7fa2af80.pub && \
-    apt-get update && \
-    apt-get -y install cuda && \
-    apt-get clean && \
-    rm -rf /var/lib/apt/lists/* && \
-    rm -rf cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb
+RUN if test ${compiler} = "nvcc-10"; then \
+      wget http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-ubuntu1804.pin && \
+      mv cuda-ubuntu1804.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
+      wget http://developer.download.nvidia.com/compute/cuda/10.1/Prod/local_installers/cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb && \
+      dpkg -i cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb && \
+      apt-key add /var/cuda-repo-10-1-local-10.1.243-418.87.00/7fa2af80.pub && \
+      apt-get update && \
+      apt-get -y install cuda-nvcc-10-1 cuda-cudart-dev-10-1 && \
+      apt-get clean && \
+      rm -rf /var/lib/apt/lists/* && \
+      rm -rf cuda-repo-ubuntu1804-10-1-local-10.1.243-418.87.00_1.0-1_amd64.deb && \
+      ln -s /usr/local/cuda-10.1 /usr/local/cuda-versioned; \
+    else \
+      wget http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-ubuntu1804.pin && \
+      mv cuda-ubuntu1804.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
+      wget http://developer.download.nvidia.com/compute/cuda/11.0.1/local_installers/cuda-repo-ubuntu1804-11-0-local_11.0.1-450.36.06-1_amd64.deb && \
+      dpkg -i cuda-repo-ubuntu1804-11-0-local_11.0.1-450.36.06-1_amd64.deb && \
+      apt-key add /var/cuda-repo-ubuntu1804-11-0-local/7fa2af80.pub && \
+      apt-get update && \
+      apt-get -y install cuda-nvcc-11-0 && \
+      apt-get clean && \
+      rm -rf /var/lib/apt/lists/* && \
+      rm -rf cuda-repo-ubuntu1804-11-0-local_11.0.1-450.36.06-1_amd64.deb && \
+      ln -s /usr/local/cuda-11.0 /usr/local/cuda-versioned; \
+    fi
 
 ENV CC=gcc \
     CXX=g++
@@ -63,7 +78,8 @@ RUN wget http://www.mpich.org/static/downloads/3.3.2/mpich-3.3.2.tar.gz && \
     cd - && \
     rm -rf mpich-3.3.2
 
-ENV CUDACXX=/usr/local/cuda-10.1/bin/nvcc
+ENV CUDACXX=/usr/local/cuda-versioned/bin/nvcc
+ENV PATH=/usr/local/cuda-versioned/bin/:$PATH
 
 RUN git clone https://github.com/kokkos/nvcc_wrapper.git && \
     cd nvcc_wrapper && \

--- a/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
+++ b/ci/docker/ubuntu-18.04-nvidia-cpp.dockerfile
@@ -60,23 +60,8 @@ RUN if test ${compiler} = "nvcc-10"; then \
 ENV CC=gcc \
     CXX=g++
 
-RUN wget http://www.mpich.org/static/downloads/3.3.2/mpich-3.3.2.tar.gz && \
-    tar xzf mpich-3.3.2.tar.gz && \
-    rm mpich-3.3.2.tar.gz && \
-    cd mpich-3.3.2 && \
-    ./configure \
-        --enable-static=false \
-        --enable-alloca=true \
-        --disable-long-double \
-        --enable-threads=single \
-        --enable-fortran=no \
-        --enable-fast=all \
-        --enable-g=none \
-        --enable-timing=none && \
-    make -j4 && \
-    make install && \
-    cd - && \
-    rm -rf mpich-3.3.2
+COPY ./ci/deps/mpich.sh mpich.sh
+RUN ./mpich.sh 3.3.2 -j4
 
 ENV CUDACXX=/usr/local/cuda-versioned/bin/nvcc
 ENV PATH=/usr/local/cuda-versioned/bin/:$PATH

--- a/cmake/turn_on_warnings.cmake
+++ b/cmake/turn_on_warnings.cmake
@@ -18,4 +18,6 @@ if(NOT hasParent)
   enable_cxx_compiler_flag_if_supported("-Wshadow")
   enable_cxx_compiler_flag_if_supported("-Wno-unknown-pragmas")
   enable_cxx_compiler_flag_if_supported("-Wsign-compare")
+  # Not really a warning, is still diagnostic related..
+  enable_cxx_compiler_flag_if_supported("-ftemplate-backtrace-limit=100")
 endif()

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -9,10 +9,12 @@
 #
 # Variables:
 #   ARCH={amd64, arm64v8, ...}
-#   COMPILER_TYPE={gnu, clang, intel}
+#   COMPILER_TYPE={gnu, clang, intel, nvidia}
 #   COMPILER={gcc-5, gcc-6, gcc-7, gcc-8, gcc-9, gcc-10,
 #             clang-3.9, clang-4.0, clang-5.0, clang-6.0, clang-7, clang-8,
-#             clang-9, clang-10, icc-18}
+#             clang-9, clang-10,
+#             icc-18, icc-19,
+#             nvcc-10}
 #   REPO=lifflander1/vt
 #   UBUNTU={18.04, 20.04}
 #   ULIMIT_CORE=0
@@ -60,6 +62,7 @@ volumes:
   amd64-ubuntu-18.04-icc-18-cache:
   amd64-ubuntu-18.04-icc-19-cache:
   amd64-ubuntu-18.04-icc-20-cache:
+  amd64-ubuntu-18.04-nvcc-10-cache:
   amd64-alpine-clang-3.9-cache:
   amd64-alpine-clang-4.0-cache:
   amd64-alpine-clang-5.0-cache:
@@ -77,6 +80,7 @@ volumes:
   amd64-alpine-icc-18-cache:
   amd64-alpine-icc-19-cache:
   amd64-alpine-icc-20-cache:
+  amd64-alpine-nvcc-10-cache:
   arm64v8-ubuntu-18.04-gcc-7-cache:
   arm64v8-alpine-gcc-7-cache:
 

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -14,7 +14,7 @@
 #             clang-3.9, clang-4.0, clang-5.0, clang-6.0, clang-7, clang-8,
 #             clang-9, clang-10,
 #             icc-18, icc-19,
-#             nvcc-10}
+#             nvcc-10, nvcc-11}
 #   REPO=lifflander1/vt
 #   UBUNTU={18.04, 20.04}
 #   ULIMIT_CORE=0
@@ -31,9 +31,24 @@
 #   BUILD_TYPE=release   # CMake build type
 #   CODE_COVERAGE=0      # Enable generation of code coverage reports
 #
-# Examples:
+# In order to run in the container, there are two stages. First, one must build
+# or pull the base container with the variables exported (or using the defaults
+# in `vt/.env` that configure it:
 #
-#  $ COMPILER=gcc-8 docker-compose build ubuntu-cpp
+# For instance, if you want an interactive intel build with icc-18 that enables
+# VT tracing, run the following commands:
+#
+# $ export COMPILER_TYPE=intel
+# $ export COMPILER=icc-18
+# $ export VT_TRACE=1
+# $ docker-compose pull ubuntu-cpp-interactive
+# $ docker-compose run ubuntu-cpp-interactive
+# # /vt/ci/build_cpp.sh /vt /build
+#
+# For a non-interactive build with gcc-8, since gnu is the default compiler
+# type, one may do the following:
+#
+#  $ COMPILER=gcc-8 docker-compose pull ubuntu-cpp
 #  $ COMPILER=gcc-8 docker-compose run ubuntu-cpp
 #
 
@@ -63,6 +78,7 @@ volumes:
   amd64-ubuntu-18.04-icc-19-cache:
   amd64-ubuntu-18.04-icc-20-cache:
   amd64-ubuntu-18.04-nvcc-10-cache:
+  amd64-ubuntu-18.04-nvcc-11-cache:
   amd64-alpine-clang-3.9-cache:
   amd64-alpine-clang-4.0-cache:
   amd64-alpine-clang-5.0-cache:
@@ -81,6 +97,7 @@ volumes:
   amd64-alpine-icc-19-cache:
   amd64-alpine-icc-20-cache:
   amd64-alpine-nvcc-10-cache:
+  amd64-alpine-nvcc-11-cache:
   arm64v8-ubuntu-18.04-gcc-7-cache:
   arm64v8-alpine-gcc-7-cache:
 

--- a/docs/md/building.md
+++ b/docs/md/building.md
@@ -113,8 +113,9 @@ The `docker` builds are configured through `docker-compose` to use a shared,
 cached filesystem mount with the host for `ccache` to enable fast re-builds.
 
 For `docker-compose`, the following variables can be set to configure the
-build. One may configure the architecture, compiler type and version, Linux
-distro (ubuntu or alpine), and distro version.
+build. One may configure the architecture, compiler type (GNU, Clang, Intel,
+NVidia) and compiler version, Linux distro (ubuntu or alpine), and distro
+version.
 
 The default set of the docker configuration options is located in `vt/.env`,
 which `docker-compose` will read.
@@ -122,10 +123,12 @@ which `docker-compose` will read.
 ```
 # Variables:
 #   ARCH={amd64, arm64v8, ...}
-#   COMPILER_TYPE={gnu, clang, intel}
+#   COMPILER_TYPE={gnu, clang, intel, nvidia}
 #   COMPILER={gcc-5, gcc-6, gcc-7, gcc-8, gcc-9, gcc-10,
 #             clang-3.9, clang-4.0, clang-5.0, clang-6.0, clang-7, clang-8,
-#             clang-9, clang-10, icc-18}
+#             clang-9, clang-10,
+#             icc-18, icc-19,
+#             nvcc-10}
 #   REPO=lifflander1/vt
 #   UBUNTU={18.04, 20.04}
 #   ULIMIT_CORE=0

--- a/docs/md/building.md
+++ b/docs/md/building.md
@@ -128,7 +128,7 @@ which `docker-compose` will read.
 #             clang-3.9, clang-4.0, clang-5.0, clang-6.0, clang-7, clang-8,
 #             clang-9, clang-10,
 #             icc-18, icc-19,
-#             nvcc-10}
+#             nvcc-10, nvcc-11}
 #   REPO=lifflander1/vt
 #   UBUNTU={18.04, 20.04}
 #   ULIMIT_CORE=0
@@ -141,10 +141,14 @@ which `docker-compose` will read.
 #   VT_TRACE_RT=0        # Enable tracing at runtime (for testing)
 #   VT_ASAN=0            # Enable address sanitizer
 #   VT_EXTENDED_TESTS=1  # Build all the extended testing
+#   VT_ZOLTAN=0          # Build with Zoltan enabled
 #   BUILD_TYPE=release   # CMake build type
+#   CODE_COVERAGE=0      # Enable generation of code coverage reports
 ```
 
-With these set, run the following for a non-interactive build with ubuntu:
+With these set, one may run the following for a non-interactive build with
+ubuntu. Or, to speed up the build process, the base container can be pulled for
+many of the common configurations: `docker-compose pull ubuntu-cpp`.
 
 ```bash
 $ cd vt

--- a/docs/md/building.md
+++ b/docs/md/building.md
@@ -114,7 +114,7 @@ cached filesystem mount with the host for `ccache` to enable fast re-builds.
 
 For `docker-compose`, the following variables can be set to configure the
 build. One may configure the architecture, compiler type (GNU, Clang, Intel,
-NVidia) and compiler version, Linux distro (ubuntu or alpine), and distro
+Nvidia) and compiler version, Linux distro (ubuntu or alpine), and distro
 version.
 
 The default set of the docker configuration options is located in `vt/.env`,

--- a/scripts/workflows.ini
+++ b/scripts/workflows.ini
@@ -59,6 +59,18 @@ vt_pool = 0
 output_name = .github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
 vt_extended_tests = "${{ github.event_name != 'pull_request' }}"
 
+[PR-tests-nvcc-10-1-extended]
+test_type = "PR tests extended"
+test_configuration = "nvidia cuda 10.1, ubuntu, mpich"
+compiler_type = nvidia
+compiler = nvcc-10
+cache_name = ubuntu-nvidia-10-cache
+vt_trace = 1
+vt_pool = 0
+output_name = .github/workflows/dockerimage-nvidia-10-ubuntu-mpich-extended.yml
+condition = "    if: github.event.label.name == 'runextendedtests'"
+triggers = "on:\\n  pull_request:\\n    types:\\n      - labeled"
+
 [PR-tests-nvcc-11-0]
 test_configuration = "nvidia cuda 11.0, ubuntu, mpich"
 compiler_type = nvidia
@@ -68,6 +80,18 @@ vt_trace = 1
 vt_pool = 0
 output_name = .github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
 vt_extended_tests = "${{ github.event_name != 'pull_request' }}"
+
+[PR-tests-nvcc-11-0-extended]
+test_type = "PR tests extended"
+test_configuration = "nvidia cuda 11.0, ubuntu, mpich"
+compiler_type = nvidia
+compiler = nvcc-11
+cache_name = ubuntu-nvidia-11-cache
+vt_trace = 1
+vt_pool = 0
+output_name = .github/workflows/dockerimage-nvidia-11-ubuntu-mpich-extended.yml
+condition = "    if: github.event.label.name == 'runextendedtests'"
+triggers = "on:\\n  pull_request:\\n    types:\\n      - labeled"
 
 [PR-tests-gcc-5]
 test_configuration = "gcc-5, ubuntu, mpich"

--- a/scripts/workflows.ini
+++ b/scripts/workflows.ini
@@ -49,6 +49,16 @@ output_name = .github/workflows/dockerimage-intel-18-ubuntu-mpich-extended.yml
 condition = "    if: github.event.label.name == 'runextendedtests'"
 triggers = "on:\\n  pull_request:\\n    types:\\n      - labeled"
 
+[PR-tests-nvcc-10-1]
+test_configuration = "nvidia cuda 10.1, ubuntu, mpich"
+compiler_type = nvidia
+compiler = nvcc-10
+cache_name = ubuntu-nvidia-10-cache
+vt_trace = 1
+vt_pool = 0
+output_name = .github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
+vt_extended_tests = "${{ github.event_name != 'pull_request' }}"
+
 [PR-tests-gcc-5]
 test_configuration = "gcc-5, ubuntu, mpich"
 compiler_type = gnu

--- a/scripts/workflows.ini
+++ b/scripts/workflows.ini
@@ -48,6 +48,7 @@ vt_pool = 0
 output_name = .github/workflows/dockerimage-intel-18-ubuntu-mpich-extended.yml
 condition = "    if: github.event.label.name == 'runextendedtests'"
 triggers = "on:\\n  pull_request:\\n    types:\\n      - labeled"
+job_name = build_optional
 
 [PR-tests-nvcc-10-1]
 test_configuration = "nvidia cuda 10.1, ubuntu, mpich"
@@ -58,6 +59,7 @@ vt_trace = 1
 vt_pool = 0
 output_name = .github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
 vt_extended_tests = "${{ github.event_name != 'pull_request' }}"
+job_name = build_optional
 
 [PR-tests-nvcc-10-1-extended]
 test_type = "PR tests extended"
@@ -70,6 +72,7 @@ vt_pool = 0
 output_name = .github/workflows/dockerimage-nvidia-10-ubuntu-mpich-extended.yml
 condition = "    if: github.event.label.name == 'runextendedtests'"
 triggers = "on:\\n  pull_request:\\n    types:\\n      - labeled"
+job_name = build_optional
 
 [PR-tests-nvcc-11-0]
 test_configuration = "nvidia cuda 11.0, ubuntu, mpich"
@@ -80,6 +83,7 @@ vt_trace = 1
 vt_pool = 0
 output_name = .github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
 vt_extended_tests = "${{ github.event_name != 'pull_request' }}"
+job_name = build_optional
 
 [PR-tests-nvcc-11-0-extended]
 test_type = "PR tests extended"
@@ -92,6 +96,7 @@ vt_pool = 0
 output_name = .github/workflows/dockerimage-nvidia-11-ubuntu-mpich-extended.yml
 condition = "    if: github.event.label.name == 'runextendedtests'"
 triggers = "on:\\n  pull_request:\\n    types:\\n      - labeled"
+job_name = build_optional
 
 [PR-tests-gcc-5]
 test_configuration = "gcc-5, ubuntu, mpich"

--- a/scripts/workflows.ini
+++ b/scripts/workflows.ini
@@ -59,6 +59,16 @@ vt_pool = 0
 output_name = .github/workflows/dockerimage-nvidia-10-ubuntu-mpich.yml
 vt_extended_tests = "${{ github.event_name != 'pull_request' }}"
 
+[PR-tests-nvcc-11-0]
+test_configuration = "nvidia cuda 11.0, ubuntu, mpich"
+compiler_type = nvidia
+compiler = nvcc-11
+cache_name = ubuntu-nvidia-11-cache
+vt_trace = 1
+vt_pool = 0
+output_name = .github/workflows/dockerimage-nvidia-11-ubuntu-mpich.yml
+vt_extended_tests = "${{ github.event_name != 'pull_request' }}"
+
 [PR-tests-gcc-5]
 test_configuration = "gcc-5, ubuntu, mpich"
 compiler_type = gnu

--- a/src/vt/messaging/active.impl.h
+++ b/src/vt/messaging/active.impl.h
@@ -55,7 +55,7 @@
 
 namespace vt { namespace messaging {
 
-constexpr ByteType msgsize_not_specified = -1;
+constexpr ByteType msgsize_not_specified = static_cast<ByteType>(-1);
 constexpr NodeType broadcast_dest = uninitialized_destination;
 
 template <typename MsgPtrT>

--- a/src/vt/messaging/message/message_serialize.h
+++ b/src/vt/messaging/message/message_serialize.h
@@ -53,10 +53,37 @@
 
 namespace vt { namespace messaging {
 
+// C++17 port. ref. https://en.cppreference.com/w/cpp/types/conjunction
+template<class...> struct cxx14_conjunction : std::true_type { };
+template<class B1> struct cxx14_conjunction<B1> : B1 { };
+template<class B1, class... Bn>
+struct cxx14_conjunction<B1, Bn...>
+  : std::conditional_t<bool(B1::value), cxx14_conjunction<Bn...>, B1> {};
+
+// C++17 port. ref. https://en.cppreference.com/w/cpp/types/void_t
+template<typename... Ts>
+struct cxx14_make_void { typedef void type;};
+template<typename... Ts>
+using cxx14_void_t = typename cxx14_make_void<Ts...>::type;
+
 // Only some compilers can directly check for a serialization method
 // because to do so requires declval on a templated member WITHOUT
 // causing templated instantiation (which will fail).
 #ifndef vt_quirked_serialize_method_detection
+
+using SerializeSizerType = ::checkpoint::Sizer;
+
+// Covering case work-about for NVCC (CUDA 10.1). See usage site.
+template <class T>
+struct has_any_serialize_member_t
+{
+  template <typename U>
+  static auto test(U* u) -> decltype(
+      u->template serialize<SerializeSizerType>(std::declval<SerializeSizerType&>())
+      , char(0)) { return char(0); }
+  static double test(...) { return 0; }
+  static const bool value = (sizeof(test((T*)0)) == 1);
+};
 
 template <typename U, typename = void>
 struct has_own_serialize_member_t : std::false_type {};
@@ -66,7 +93,7 @@ struct has_own_serialize_member_t<U,
   std::enable_if_t<
     std::is_same<
       void (U::*)(::checkpoint::Sizer&),
-      decltype(&U::template serialize<::checkpoint::Sizer&>)
+      decltype(&U::template serialize<SerializeSizerType>)
     >::value
   >>
   : std::true_type
@@ -77,7 +104,14 @@ struct has_own_serialize_member_t<U,
 template <typename T>
 static constexpr auto const has_own_serialize =
   ::checkpoint::SerializableTraits<T>::has_serialize_noninstrusive
-  or has_own_serialize_member_t<T>::value;
+  // NVCC (CUDA 10.1 and 11) has SFINAE issues with decltype above,
+  // and possible fallout with lazy evaluation of 'and'.
+  // While this compiles, NVCC incorrectly detects some types
+  // as requiring serialization. See #949.
+  or cxx14_conjunction<
+    has_any_serialize_member_t<T>,
+    has_own_serialize_member_t<T>
+  >::value;
 
 #endif
 
@@ -222,19 +256,6 @@ struct is_byte_copyable_t {
   constexpr static bool value = std::is_trivially_destructible<T>::value and not std::is_pointer<T>::value;
 };
 #endif
-
-// C++17 port. ref. https://en.cppreference.com/w/cpp/types/conjunction
-template<class...> struct cxx14_conjunction : std::true_type { };
-template<class B1> struct cxx14_conjunction<B1> : B1 { };
-template<class B1, class... Bn>
-struct cxx14_conjunction<B1, Bn...>
-  : std::conditional_t<bool(B1::value), cxx14_conjunction<Bn...>, B1> {};
-
-// C++17 port. ref. https://en.cppreference.com/w/cpp/types/void_t
-template<typename... Ts>
-struct cxx14_make_void { typedef void type;};
-template<typename... Ts>
-using cxx14_void_t = typename cxx14_make_void<Ts...>::type;
 
 // Foward-declare
 struct BaseMsg;

--- a/src/vt/rdmahandle/handle.index.impl.h
+++ b/src/vt/rdmahandle/handle.index.impl.h
@@ -49,14 +49,10 @@
 
 namespace vt { namespace rdma {
 
-/**
- * \brief Convenience using for when U is a vt::NodeType
- */
-template <typename U>
-using isIdx = typename std::enable_if_t<not std::is_same<U,vt::NodeType>::value>;
-
 template <typename T, HandleEnum E, typename I>
-bool Handle<T,E,I,isIdx<I>>::ready() const {
+bool Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::ready() const {
   if (proxy_ != no_obj_group) {
     auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
     return proxy.get()->ready();
@@ -66,7 +62,9 @@ bool Handle<T,E,I,isIdx<I>>::ready() const {
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::readExclusive(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::readExclusive(
   std::function<void(T const*, std::size_t count)> fn
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -74,7 +72,9 @@ void Handle<T,E,I,isIdx<I>>::readExclusive(
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::readShared(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::readShared(
   std::function<void(T const*, std::size_t count)> fn
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -82,7 +82,9 @@ void Handle<T,E,I,isIdx<I>>::readShared(
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::modifyExclusive(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::modifyExclusive(
   std::function<void(T*, std::size_t count)> fn
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -90,7 +92,9 @@ void Handle<T,E,I,isIdx<I>>::modifyExclusive(
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::modifyShared(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::modifyShared(
   std::function<void(T*, std::size_t count)> fn
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -98,17 +102,23 @@ void Handle<T,E,I,isIdx<I>>::modifyShared(
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::lock(Lock l, I const& index) {
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::lock(Lock l, I const& index) {
   //@todo: implement this
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::unlock() {
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::unlock() {
   this->lock_ = nullptr;
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::get(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::get(
   I const& index, T* ptr, std::size_t len, int offset, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -116,8 +126,12 @@ void Handle<T,E,I,isIdx<I>>::get(
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isIdx<I>>::RequestType
-Handle<T,E,I,isIdx<I>>::rget(
+typename Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::rget(
   I const& index, T* ptr, std::size_t len, int offset, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -125,7 +139,9 @@ Handle<T,E,I,isIdx<I>>::rget(
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::get(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::get(
   I const& index, std::size_t len, int offset, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -133,8 +149,12 @@ void Handle<T,E,I,isIdx<I>>::get(
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isIdx<I>>::RequestType
-Handle<T,E,I,isIdx<I>>::rget(
+typename Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::rget(
   I const& index, std::size_t len, int offset, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -163,7 +183,9 @@ Handle<T,E,I,isIdx<I>>::rget(
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::put(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::put(
   I const& index, T* ptr, std::size_t len, int offset, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -171,8 +193,12 @@ void Handle<T,E,I,isIdx<I>>::put(
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isIdx<I>>::RequestType
-Handle<T,E,I,isIdx<I>>::rput(
+typename Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::rput(
   I const& index, T* ptr, std::size_t len, int offset, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -180,7 +206,9 @@ Handle<T,E,I,isIdx<I>>::rput(
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isIdx<I>>::accum(
+void Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::accum(
   I const& index, T* ptr, std::size_t len, int offset, MPI_Op op, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -188,8 +216,12 @@ void Handle<T,E,I,isIdx<I>>::accum(
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isIdx<I>>::RequestType
-Handle<T,E,I,isIdx<I>>::raccum(
+typename Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::raccum(
   I const& index, T* ptr, std::size_t len, int offset, MPI_Op op, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -197,7 +229,9 @@ Handle<T,E,I,isIdx<I>>::raccum(
 }
 
 template <typename T, HandleEnum E, typename I>
-T Handle<T,E,I,isIdx<I>>::fetchOp(
+T Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::fetchOp(
   I const& index, T ptr, int offset, MPI_Op op, Lock l
 ) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
@@ -205,7 +239,9 @@ T Handle<T,E,I,isIdx<I>>::fetchOp(
 }
 
 template <typename T, HandleEnum E, typename I>
-std::size_t Handle<T,E,I,isIdx<I>>::getCount(I const& index) {
+std::size_t Handle<
+  T,E,I,typename std::enable_if_t<not std::is_same<I,vt::NodeType>::value>
+>::getCount(I const& index) {
   auto proxy = vt::objgroup::proxy::Proxy<SubHandle<T,E,I>>(proxy_);
   if (proxy.get()->isUniform()) {
     return this->count();

--- a/src/vt/rdmahandle/handle.node.impl.h
+++ b/src/vt/rdmahandle/handle.node.impl.h
@@ -49,42 +49,50 @@
 
 namespace vt { namespace rdma {
 
-/**
- * \brief Convenience using for when U is a vt::NodeType
- */
-template <typename U>
-using isNode = typename std::enable_if_t<std::is_same<U,vt::NodeType>::value>;
-
 template <typename T, HandleEnum E, typename I>
-bool Handle<T,E,I,isNode<I>>::ready() const {
+bool Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::ready() const {
   return vt::theHandleRDMA()->getEntry<T,E>(key_).ready();
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::get(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::get(
   vt::NodeType node, T* ptr, std::size_t len, int offset, Lock l
 ) {
   return vt::theHandleRDMA()->getEntry<T,E>(key_).get(node, l, ptr, len, offset + this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isNode<I>>::RequestType
-Handle<T,E,I,isNode<I>>::rget(
+typename Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::rget(
   vt::NodeType node, T* ptr, std::size_t len, int offset, Lock l
 ) {
     return vt::theHandleRDMA()->getEntry<T,E>(key_).rget(node, l, ptr, len, offset + this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::get(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::get(
   vt::NodeType node, std::size_t len, int offset, Lock l
 ) {
   rget(node, len, offset);
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isNode<I>>::RequestType
-Handle<T,E,I,isNode<I>>::rget(vt::NodeType node, std::size_t len, int offset, Lock l) {
+typename Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::rget(vt::NodeType node, std::size_t len, int offset, Lock l) {
   if (this->getBuffer() == nullptr) {
     auto ptr = std::make_unique<T[]>(len);
     auto r = vt::theHandleRDMA()->getEntry<T,E>(key_).rget(
@@ -110,44 +118,58 @@ Handle<T,E,I,isNode<I>>::rget(vt::NodeType node, std::size_t len, int offset, Lo
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::put(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::put(
   vt::NodeType node, T* ptr, std::size_t len, int offset, Lock l
 ) {
   return vt::theHandleRDMA()->getEntry<T,E>(key_).put(node, l, ptr, len, offset + this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isNode<I>>::RequestType
-Handle<T,E,I,isNode<I>>::rput(
+typename Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>>::rput(
   vt::NodeType node, T* ptr, std::size_t len, int offset, Lock l
 ) {
   return vt::theHandleRDMA()->getEntry<T,E>(key_).rput(node, l, ptr, len, offset + this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::accum(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::accum(
   vt::NodeType node, T* ptr, std::size_t len, int offset, MPI_Op op, Lock l
 ) {
   return vt::theHandleRDMA()->getEntry<T,E>(key_).accum(node, l, ptr, len, offset + this->hoff(), op);
 }
 
 template <typename T, HandleEnum E, typename I>
-typename Handle<T,E,I,isNode<I>>::RequestType
-Handle<T,E,I,isNode<I>>::raccum(
+typename Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::RequestType
+Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::raccum(
   vt::NodeType node, T* ptr, std::size_t len, int offset, MPI_Op op, Lock l
 ) {
   return vt::theHandleRDMA()->getEntry<T,E>(key_).raccum(node, l, ptr, len, offset + this->hoff(), op);
 }
 
 template <typename T, HandleEnum E, typename I>
-T Handle<T,E,I,isNode<I>>::fetchOp(
+T Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::fetchOp(
   vt::NodeType node, T ptr, int offset, MPI_Op op, Lock l
 ) {
   return vt::theHandleRDMA()->getEntry<T,E>(key_).fetchOp(node, l, ptr, offset, op);
 }
 
 template <typename T, HandleEnum E, typename I>
-std::size_t Handle<T,E,I,isNode<I>>::getCount(vt::NodeType node) {
+std::size_t Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::getCount(vt::NodeType node) {
   if (vt::theHandleRDMA()->getEntry<T,E>(key_).isUniform()) {
     return this->count();
   } else {
@@ -156,72 +178,96 @@ std::size_t Handle<T,E,I,isNode<I>>::getCount(vt::NodeType node) {
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::readExclusive(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::readExclusive(
   std::function<void(T const*, std::size_t count)> fn
 ) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).access(Lock::Exclusive, fn, this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::readShared(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::readShared(
   std::function<void(T const*, std::size_t count)> fn
 ) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).access(Lock::Shared, fn, this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::modifyExclusive(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::modifyExclusive(
   std::function<void(T*, std::size_t count)> fn
 ) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).access(Lock::Exclusive, fn, this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::modifyShared(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::modifyShared(
   std::function<void(T*, std::size_t count)> fn
 ) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).access(Lock::Shared, fn, this->hoff());
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::access(
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::access(
   Lock l, std::function<void(T*, std::size_t count)> fn, std::size_t offset
 ) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).access(l, fn, this->hoff() +  offset);
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::fence(int assert) {
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::fence(int assert) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).fence(assert);
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::sync() {
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::sync() {
   vt::theHandleRDMA()->getEntry<T,E>(key_).sync();
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::flush(vt::NodeType node) {
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::flush(vt::NodeType node) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).flush(node);
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::flushLocal(vt::NodeType node) {
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::flushLocal(vt::NodeType node) {
   vt::theHandleRDMA()->getEntry<T,E>(key_).flushLocal(node);
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::flushAll() {
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::flushAll() {
   vt::theHandleRDMA()->getEntry<T,E>(key_).flushAll();
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::lock(Lock l, vt::NodeType node) {
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::lock(Lock l, vt::NodeType node) {
   this->lock_ = vt::theHandleRDMA()->getEntry<T,E>(key_).lock(l, node);
 }
 
 template <typename T, HandleEnum E, typename I>
-void Handle<T,E,I,isNode<I>>::unlock() {
+void Handle<
+  T,E,I,typename std::enable_if_t<std::is_same<I,vt::NodeType>::value>
+>::unlock() {
   this->lock_ = nullptr;
 }
 

--- a/src/vt/rdmahandle/manager.impl.h
+++ b/src/vt/rdmahandle/manager.impl.h
@@ -53,6 +53,9 @@ namespace vt { namespace rdma {
 namespace impl {
 
 struct HandleData {
+  // #949. NVCC shim: Ensure type is compatible with serialization.
+  using isByteCopyable = std::true_type;
+
   HandleData();
   HandleData(HandleKey in_key, std::size_t in_size, int in_count)
     : key_(in_key),

--- a/tests/perf/ping_pong.cc
+++ b/tests/perf/ping_pong.cc
@@ -105,7 +105,7 @@ static void finishedPing(FinishedPingMsg<num_bytes>* msg) {
 
   if (num_bytes != max_bytes) {
     auto pmsg = makeMessage<PingMsg<num_bytes * 2>>();
-    theMsg()->sendMsg<PingMsg<num_bytes * 2>, pingPong>(
+    theMsg()->sendMsg<PingMsg<num_bytes * 2>, pingPong<num_bytes * 2>>(
       pong_node, pmsg.get()
     );
   }
@@ -133,7 +133,7 @@ static void pingPong(PingMsg<num_bytes>* in_msg) {
 
   if (cnt >= num_pings) {
     auto msg = makeMessage<FinishedPingMsg<num_bytes>>(num_bytes);
-    theMsg()->sendMsg<FinishedPingMsg<num_bytes>, finishedPing>(
+    theMsg()->sendMsg<FinishedPingMsg<num_bytes>, finishedPing<num_bytes>>(
       0, msg.get()
     );
   } else {
@@ -141,12 +141,12 @@ static void pingPong(PingMsg<num_bytes>* in_msg) {
       theContext()->getNode() == ping_node ? pong_node : ping_node;
     #if REUSE_MESSAGE_PING_PONG
       // @todo: fix this memory allocation problem
-      theMsg()->sendMsg<PingMsg<num_bytes>, pingPong>(
+      theMsg()->sendMsg<PingMsg<num_bytes>, pingPong<num_bytes>>(
         next, in_msg, [=]{ /*delete in_msg;*/ }
       );
     #else
       auto m = makeMessage<PingMsg<num_bytes>>(cnt + 1);
-      theMsg()->sendMsg<PingMsg<num_bytes>, pingPong>(next, m.get());
+      theMsg()->sendMsg<PingMsg<num_bytes>, pingPong<num_bytes>>(next, m.get());
     #endif
   }
 }
@@ -169,7 +169,7 @@ int main(int argc, char** argv) {
 
   if (my_node == 0) {
     auto m = makeMessage<PingMsg<min_bytes>>();
-    theMsg()->sendMsg<PingMsg<min_bytes>, pingPong>(pong_node, m.get());
+    theMsg()->sendMsg<PingMsg<min_bytes>, pingPong<min_bytes>>(pong_node, m.get());
   }
 
   while (!rt->isTerminated()) {

--- a/tests/unit/pool/test_pool_message_sizes.cc
+++ b/tests/unit/pool/test_pool_message_sizes.cc
@@ -102,12 +102,12 @@ void TestPoolMessageSizes::testPoolFun(TestMsg<num_bytes>* prev_msg) {
 
   if (count < max_test_count) {
     auto msg = makeMessage<TestMsg<num_bytes>>();
-    theMsg()->sendMsg<TestMsg<num_bytes>, testPoolFun>(
+    theMsg()->sendMsg<TestMsg<num_bytes>, testPoolFun<num_bytes>>(
       next, msg.get()
     );
   } else {
     auto msg = makeMessage<TestMsg<num_bytes * 2>>();
-    theMsg()->sendMsg<TestMsg<num_bytes * 2>, testPoolFun>(
+    theMsg()->sendMsg<TestMsg<num_bytes * 2>, testPoolFun<num_bytes * 2>>(
       next, msg.get()
     );
     count = 0;

